Allow for switching between emoji categories using swipe left/right

eranl opened the pull request:
This is a naive implementation, with no swipe animation. Is it good enough?

Fixes #1426. Fixes #1332.

Helium314 commented:
>This is a naive implementation, with no swipe animation. Is it good enough?

I expect without animations there will soon be an issue about it, but I'd say the current PR still does what is requested, and probably a lot of people will still be fine without animations.

Animations possibly require considerably more work than what you did. Maybe a `ViewPager2` (successor of the `TabWidget` that is commented in `emojiPalettesView`) could help.

eranl commented:
> Animations possibly require considerably more work than what you did. Maybe a `ViewPager2` (successor of the `TabWidget` that is commented in `emojiPalettesView`) could help.

Let me give that a try.

eranl commented:
Ok, switched to `ViewPager2`. I think it looks better. 🙂 

I don't understand what the four unused `Indicator`-related fields in `EmojiPalettesView` were used for, and whether I'm supposed to use them for this somehow. Or can they be deleted?

I guess the methods I added to `EmojiCategory` should be deleted, right?

Helium314 commented:
> Ok, switched to `ViewPager2`. I think it looks better. 🙂

Definitely.
A few minor things I noticed:
* when opening the emoji view for the first time, there is a scroll animation if you don't start on the recents page, which is a little irritating
* sometimes when switching tabs using swipe, the `mEmojiCategoryPageIndicatorView` is not updated until you scroll up/down
* sometimes when switching tabs using swipe, you lose the scroll offset
* swiping shows the same emoji category coming in again, until you release the finger to trigger the actual switch

> I don't understand what the four unused `Indicator`-related fields in `EmojiPalettesView` were used for, and whether I'm supposed to use them for this somehow. Or can they be deleted?

They are remains from using the `TabWidget` which I decided to remove a while ago (not sure why though). In Holo theme there was a small indicator icon below the selected emoji.
I actually thought about switching to ViewPager2 initially, but didn't find enough time / priority to do so. That's why I had simply left some commented code from the old style, but no one complained about the missing indicator in over a year.

> I guess the methods I added to `EmojiCategory` should be deleted, right?

They aren't used any more, so better remove them.

Helium314 commented on app/src/main/res/layout/emoji_category_view.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+    Copyright (C) 2013 The Android Open Source Project
```
Where did you get this ancient File?

eranl replied:
Copied `emoji_palettes_view.xml`

Helium314 replied:
Ah, I was quite confused about where the copyright notice came from.

eranl commented:
> * when opening the emoji view for the first time, there is a scroll animation if you don't start on the recents page, which is a little irritating

Fixed.

> * sometimes when switching tabs using swipe, the `mEmojiCategoryPageIndicatorView` is not updated until you scroll up/down

Fixed.

> * sometimes when switching tabs using swipe, you lose the scroll offset

This seems to be the effect of view recycling. I think we can prevent recycling by setting it to keep all tabs in memory. Should I try that?

> * swiping shows the same emoji category coming in again, until you release the finger to trigger the actual switch

I don't see that. Did it get fixed for you?

> > I don't understand what the four unused `Indicator`-related fields in `EmojiPalettesView` were used for, and whether I'm supposed to use them for this somehow. Or can they be deleted?
> 
> They are remains from using the `TabWidget` which I decided to remove a while ago (not sure why though). In Holo theme there was a small indicator icon below the selected emoji. I actually thought about switching to ViewPager2 initially, but didn't find enough time / priority to do so. That's why I had simply left some commented code from the old style, but no one complained about the missing indicator in over a year.

Should I restore that indicator? Remove these fields?

> > I guess the methods I added to `EmojiCategory` should be deleted, right?
> 
> They aren't used any more, so better remove them.

Done.

I've read that `RecyclerView` can be made to work like `ViewPager2` rather easily. Should I try that, in order to avoid the extra weight?

Should I make the anonymous `RecyclerView.Adapter` class that I added a named inner class, or a top-level class?

Helium314 commented:
> I don't see that. Did it get fixed for you?

Yes fixed, thanks! Looks much better now.

> Should I restore that indicator? Remove these fields?

Just leave it as it is, I'd like to have the cleanup separate from this PR, even if it's somewhat connected.
No one missed the indicator, and in my opinion the coloring in the top bar is enough indication for which tab is active.

> >  sometimes when switching tabs using swipe, you lose the scroll offset
>
> This seems to be the effect of view recycling. I think we can prevent recycling by setting it to keep all tabs in memory. Should I try that?

Sorry, that was my mistake. I had wrongly thought the scroll state would be more permanent, but it's not. Looks like the behavior is not affected by your PR.

> I've read that RecyclerView can be made to work like ViewPager2 rather easily. Should I try that, in order to avoid the extra weight?

If you want to try / learn / "play", you're welcome to.
The ViewPager2 appears to add 15 kB to APK size, so in my opinion the savings alone are not enough to justify extra work.

> Should I make the anonymous RecyclerView.Adapter class that I added a named inner class, or a top-level class?

I think an inner class makes sense. It shouldn't be much work compared to the current state.

eranl commented:
> Looks like the behavior is not affected by your PR.

Before this change, there was a single `RecyclerView`, so there was no way to preserve scroll state across categories (sans saving it ourselves). Now, every category has its own `RecyclerView` modulo recycling, so scroll state *is* preserved in adjacent categories, and I think we can have it preserved across all.

> The ViewPager2 appears to add 15 kB to APK size, so in my opinion the savings alone are not enough to justify extra work.

Thanks. I'll leave it as `ViewPager2` then.

> I think an inner class makes sense. 

Done.

Helium314 commented:
> I think we can have it preserved across all.

I think preserving scroll state would be expected behavior, but I only rarely use the emoji view so I might be mistaken.

eranl commented:
While testing, I encountered instability when making rapid two-dimensional scrolling and tabbing, and realized it was due to race conditions around the shared `EmojiPalettesAdapter`, which relied on `EmojiCategory`'s `getCurrent*` methods, so I changed `EmojiPalettesAdapter` to be instance-per-page.

While making that change, I noticed that the code using `mActivePosition` seems broken, since it's always `==0`. Also, I don't understand the purpose of `mActiveKeyboardViews`, and whether it was only needed because `EmojiPalettesAdapter` served all pages. Some code uses scroll positions as its keys, and other code uses `mEmojiCategory.getRecentTabId()` as key.

I left some commented out code in both `EmojiPalettesAdapter` and `EmojiPalettesView` (to where I moved two of `EmojiPalettesAdapter`'s methods) in order to check with you.

Helium314 commented:
You're seeing some of the emoji weirdness I've wondered about, but never nearly found the time to try understanding what's going on. It always seemed unnecessarily complicated to me, even more than average for this app.

I don't think I can be of much help when it comes to understanding the emoji keyboard system. Though I vaguely remember there is a `DynamicGridKeyboard` and / or `EmojiPageKeyboardView` for every 3 lines of emojis. This is likely related to `mActiveKeyboardViews`, which might be holding the mulitple keyboard views per tab?
The code actually using `mActivePosition` has been commented since OpenBoard times.

eranl commented:
The thing is `releaseCurrentKey` still uses `mActivePosition`, but it's obviously broken.

It looks like what this code tried to achieve is calling `releaseCurrentKey` & `deallocateMemory` when a `KeyboardView` is removed from view, so I replaced it with an `onViewDetachedFromWindow` handler.

It seems to be working fine to me. Any specific testing you recommend?

Helium314 commented:
There seems to be some issue in combination with enabling / disabling / adjusting one-handed mode, where emoji views should adjust size. Similar for split keyboard.

And after playing with this and scrolling I got
```
Process: helium314.keyboard.debug, PID: 6895
java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid item position 7(offset:7).state:9 androidx.recyclerview.widget.RecyclerView{f2dc09d VFED..... .......D 0,0-720,327 #7f0a0071 app:id/emoji_keyboard_list}, adapter:helium314.keyboard.keyboard.emoji.EmojiPalettesAdapter@df16191, layout:androidx.recyclerview.widget.LinearLayoutManager@d990912, context:android.view.ContextThemeWrapper@d888e3
at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6917)
at androidx.recyclerview.widget.GapWorker.prefetchPositionWithDeadline(GapWorker.java:294)
at androidx.recyclerview.widget.GapWorker.flushTaskWithDeadline(GapWorker.java:354)
at androidx.recyclerview.widget.GapWorker.flushTasksWithDeadline(GapWorker.java:370)
at androidx.recyclerview.widget.GapWorker.prefetch(GapWorker.java:377)
at androidx.recyclerview.widget.GapWorker.run(GapWorker.java:408)
at android.os.Handler.handleCallback(Handler.java:883)
at android.os.Handler.dispatchMessage(Handler.java:100)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:7356)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:491)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:940)
```
Couldn't reproduce it yet, but I suspect that `EmojiPalettesView.clearKeyboardCache` doesn't result in recreation of the keyboards any more.

eranl commented:
> There seems to be some issue in combination with enabling / disabling / adjusting one-handed mode, where emoji views should adjust size. Similar for split keyboard.

Fixed.

Helium314 commented:
Clearing the keyboard cache on close made me have a look at how many `EmojiPageKeyboardView`s and `DynamicGridKeyboard`s are created:

| | old | new |
| ----- | ----- | ----- |
| open emojis (2nd tab) | 11 keyboards, 2 views | 11 keyboards, 2 views |
| scroll down | 4 views | 4 views |
| switch to 3rd tab | 15 keyboards | 30 keyboards, 5 views |
| scroll down | 1 view | 5 views |
| close emoji keyboard and open again (2nd tab) | 2 views | 10 keyboards, 2 views |
| scroll down | 4 views | 1 keyboard, 5 views |
| switch to 3rd tab | - | 30 keyboards, 7 views |
| scroll down | 1 view | 3 views |

Creating new views on switching tab is unavoidable, so no issue with that.
But I think it would be good to avoid creating so many views keyboards if possible. Especially the number of keyboards seems strange. It's not just that the cache is dropped on exiting emoji view, but also twice as many keyboard created for the 3rd tab.

eranl commented:
Thanks for testing. I'll look into that.

I'm guessing that switching back and forth between tabs would result in fewer views created, since they should remain in memory.
Also, the fact that the cache is not cleared today is probably the reason for bugs like https://github.com/Helium314/HeliBoard/issues/1332, which gets fixed here.

Helium314 commented:
Bugs like #1332 could get fixed a lot faster if people would start using the template. I'm so tired of seeing super lazy issues with very little information, I often just unsubscribe right after opening them.

But if clearing the cache solves the issue, then we can just clear it when switching split keyboard mode, instead of every time closing the emoji view?

eranl commented:
> But if clearing the cache solves the issue, then we can just clear it when switching split keyboard mode, instead of every time closing the emoji view?

Actually, I just noticed commit 46f92276159134660069f3af194bd43908940562. Did it fix #1332?

Helium314 commented:
I don't think so, there are still often recent emojis missing in split mode

eranl commented:
> I don't think so, there are still often recent emojis missing in split mode

Oh, I think it might actually be a result of the bug I fixed [here](https://github.com/Helium314/HeliBoard/pull/1488/commits/d557f0facd0c2e2b30118ed26dea94c96892b8cd). Should I fix it in a separate PR?

eranl commented:
> Clearing the keyboard cache on close

Turns out this was unnecessary. Removed.

eranl commented:
> twice as many keyboard created for the 3rd tab.

I don't see that. I see 15 keyboards created - all keyboards for the category. It could be made more efficient by creating keyboards on demand.

Helium314 commented:
> Oh, I think it might actually be a result of the bug I fixed [here](https://github.com/Helium314/HeliBoard/pull/1488/commits/d557f0facd0c2e2b30118ed26dea94c96892b8cd). Should I fix it in a separate PR?

I think the accidental fix fit very well in this PR.

> I don't see that. I see 15 keyboards created - all keyboards for the category. It could be made more efficient by creating keyboards on demand.

Strange, I'll check again what's going on on my phone. Anyway, not clearing the cache seems more relevant to me, and that's done.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/KeyboardSwitcher.java:
```diff
@@ -638,7 +638,7 @@ public void deallocateMemory() {
         }
         if (mEmojiPalettesView != null) {
-            mEmojiPalettesView.stopEmojiPalettes();
+            mEmojiPalettesView.clearKeyboardCache();
```
`deallocateMemory` is called 10 seconds after the keyboard is out of view. I think clearing emoji keyboard cache here is too aggressive.
If you want to release memory in case the system is low on memory, `onTrimMemory` is a better place.

eranl replied:
Done.

Helium314 replied:
I re-added `mEmojiPalettesView.stopEmojiPalettes();`. I'm not sure it's necessary, but I'd rather keep the old behavior if there is no necessity to remove it.
